lib/t_web/socket/socket.ex

```diff
@@ -0,0 +1,103 @@
+defmodule TWeb.Socket do
+  @moduledoc false
+  import TWeb.Socket.ErrorHelpers
+  @behaviour :cowboy_websocket
+  require Logger
+
+  @type assigns :: term
+
+  @callback connect(:cowbot.req()) :: {:ok, assigns} | {:error, :cowboy.status_code()}
+  @callback init(assigns :: map) :: {:ok, assigns}
+
+  @callback handle_event(event :: String.t(), params :: term, assigns) ::
+              {:ok, assigns}
+              | {:ok, term, assigns}
+              | {:error, [pos_integer() | String.t()], assigns}
+
+  @callback handle_info(message :: term, assigns) ::
+              {pushes :: [String.t() | term], assigns}
+
+  @callback terminate(:cowboy_websocket.terminate_reason(), assigns) :: any
+
+  @impl true
+  def init(req, [handler] = state) do
+    # TODO hibernate after 15 secs
+    # TODO check origin?
+    case handler.connect(req) do
+      {:ok, assigns} ->
+        {:cowboy_websocket, req, [handler | assigns], _opts = %{}}
+
+      {:error, status_code} ->
+        # TODO
+        {:ok, :cowboy_req.reply(status_code, req), state}
```

Comment:
this might not work with urlsessionwebsockettask since it doesn't return the status code in the error so we won't be able to distinguish between 403, 418, and actually invalid server response since all three get the same error in urlsessionwebsockettask